sklearn/utils/validation.py

```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
 
     if max_val is not None and x > max_val:
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
+                         order=None, **kwargs):
+    """Validate sample weights
+
+    Parameters
+    ----------
+    sample_weight : {ndarray, Number or None}
+       input sample weights
+    y: ndarray or None
+       target variable. Either y or n_samples must be provided.
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
+       dtype of the validated sample_weight
+    order : 'F', 'C' or None (default=None)
+        Whether an array will be forced to be fortran or c-style.
+        When order is None (default), if ``sample_weights`` is an ndarray,
+        nothing is ensured about the memory layout of the output array,
+        otherwise it will be of 'C' order by default.
+
+    kwargs:
+       additional parameters to pass to check_array
+
+    Parameters
+    ----------
+    sample_weight : ndarray, shape=(n_samples,)
+       validated sample weights
+    """
+    if n_samples is not None and y is not None:
+        raise ValueError('Only one of y, n_samples must be provided!')
+    elif y is not None:
+        n_samples = y.shape[0]
+
+    if sample_weight is None or isinstance(sample_weight, numbers.Number):
+        if order is None:
+            order = 'C'
+        if sample_weight is None:
+            sample_weight = np.ones(n_samples, dtype=dtype, order=order)
+        else:
+            sample_weight = np.full(n_samples, sample_weight,
+                                    dtype=dtype, order=order)
+    else:
+        if dtype is None:
+            dtype = [np.float64, np.float32]
+        sample_weight = check_array(
+                sample_weight, accept_sparse=False,
+                ensure_2d=False, dtype=dtype, order=order, **kwargs
+        )
+        if sample_weight.ndim != 1:
+            raise ValueError("Sample weights must be 1D array or scalar")
+
+        if sample_weight.shape != (n_samples,):
```

Comment:
you can use `check_consistency_length`

Reply:
to have consistent error raised

Reply:
Yes, but `y` is not always provided, and `check_consistency_length` wouldn't work when only `n_samples` is provided.

Reply:
uhm True. It is a pity that we cannot keep the error message in one location.

Reply:
unless you check against X :)

Reply:
You are right, checking against `X` is probably best. Will add it.

Reply:
Was this addressed?

Reply:
No. I just want to check the shape, while `check_consistent_length` does a bunch of generic and more complicated stuff. I'm also not convinced that the resulting generic error message,
```
Found input variables with inconsistent numbers of samples: %r
```
is an improvement over the current specific message,
```
samples_weight.shape == {}, expected {}!
```
(though I'm open to suggestions on how to improve the latter one).

Reply:
The difference is also that by this point `sample_weight` has been validated, so I am sure it's an array, while `check_consistent_length` doesn't make any assumptions hence the more complicated and generic solution.

Reply:
Makes sense.